### PR TITLE
Profile guided compilation fixes

### PR DESCRIPTION
**Pull request by davidwrighton**
- Fix issue where if we can't load a token during profile data resolution, we crash with a NullReferenceException
- 
- Fix issue where when profile specified methods are added to the compilation, they cause methods not specified to be compiled as well. This caused about a 3X bloat in method compilation when working off of profile specified data
  - Fixed by adding a dependency phase model to dependency analyzer
  - Then using that to tie into method compilation and dependencies so that some extra dependencies are trimmed by our dependency analysis flow

### CODE REVIEW

**Comment by davidwrighton** [comment]
@dotnet/crossgen-contrib

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/MethodWithGCInfo.cs [hunk 1]
```diff
@@ -34,6 +35,20 @@ public MethodWithGCInfo(MethodDesc methodDesc)
             _method = methodDesc;
         }
 
+        protected override void OnMarked(NodeFactory context)
+        {
+            // Once past phase 1, no new methods which are interesting for compilation may be marked except for methods
```
Comment (trylek) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/MethodWithGCInfo.cs:40]: Would there be a chance to validate some of these claims via assertions or some such?

**Review by trylek — APPROVED** [review]
Overall this looks great to me but I'd be much happier if @MichalStrehovsky could also throw an eye on this as I believe he was one of the original architects of the CoreRT dependency analysis.

File: src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/DeferredTillPhaseNode.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (MichalStrehovsky) [inline src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/DeferredTillPhaseNode.cs:9]: Nit: unnecessary usings

**Review by MichalStrehovsky — APPROVED** [review]

**Pull request merged**